Add cookie-based sessions

connorjayr opened the pull request:
Uses the `actix-session` crate to create rudimentary cookie-based sessions. Eventually, we will want to store the session data in persistent, server-side storage such as Redis.

There is also a "functional" login page, which includes a request-response transaction with the server. However, the server does not check that the credentials are valid (in fact, credentials are not even stored yet).

connorjayr committed:
Add cookie-based sessions

Uses the `actix-session` crate to create rudimentary cookie-based sessions. Eventually, we will want to store the session data in persistent, server-side storage such as Redis.

There is also a "functional" login page, which includes a request-response transaction with the server. However, the server does not check that the credentials are valid (in fact, credentials are not even stored yet).

jputlock commented on server/src/main.rs:
```diff
@@ -1,19 +1,46 @@
+/// 32 randomly-generated bytes.
+fn get_session_secret() -> Key {
+    let session_secret =
+        env::var("SESSION_SECRET").expect("environment variable `SESSION_SECRET` must be defined");
```
The `dotenv` crate provides a `var` function that we can use here, so we don't need to `use std::env`

jputlock replied:
This also guarantees that the `.env` is loaded at the time we look up the value.

connorjayr replied:
Changed in https://github.com/connorjayr/sust/pull/3/commits/807aec22384c91cb99e3b126cec730ae560bc5b4.

jputlock commented on server/src/main.rs:
```diff
@@ -1,19 +1,46 @@
+                    .build(),
+            )
+            .service(
+                web::scope("/api")
```
Why `/api`?

connorjayr replied:
I'm planning for all back-end server routes to be prefixed with `/api`. This is in contrast to all SPA routes, which will just contain the user-friendly name (e.g. `/login`, which is handled by the React router). This makes it clear which routes are handled by the front end and which are handled by the back end.

connorjayr replied:
API (back-end) routes will also interface using `Content-Type: application/json`, rather than serve HTML content. These routes are not meant to be viewable by the user, they are only meant for the front end to make `fetch` calls in the background.

jputlock reviewed (dismissed):
Left a few comments; otherwise seems like a good start to me.

connorjayr committed:
Use dotenv crate to fetch environment variables

The `dotenv` crate provides a function that returns the value of an environment variable at runtime. Using this function, we no longer need to call `dotenv().ok()` in .
